packages/consensus/src/types/utxo_set.cairo

```diff
@@ -13,60 +13,131 @@ use core::dict::Felt252Dict;
 use core::hash::{HashStateTrait, HashStateExTrait};
 use core::poseidon::PoseidonTrait;
 use super::transaction::OutPoint;
-use super::utreexo::{UtreexoState, UtreexoAccumulator};
 
 pub const TX_OUTPUT_STATUS_NONE: u8 = 0;
 pub const TX_OUTPUT_STATUS_UNSPENT: u8 = 1;
 pub const TX_OUTPUT_STATUS_SPENT: u8 = 2;
 
 #[derive(Default, Destruct)]
 pub struct UtxoSet {
-    /// Utreexo state.
-    pub utreexo_state: UtreexoState,
-    /// The leaves represent the poseidon hashes of a block's outpoints, i.e. utxos
+    /// Hashes of created UTXOs.
     pub leaves_to_add: Array<felt252>,
-    /// Hashes of UTXOs created within the current block(s).
-    /// Note that to preserve the ordering, cache has to be updated right after a
+    /// Hashes of spent (i.e. deleted) UTXOs.
+    pub leaves_to_delete: Array<felt252>,
+    /// Number of pending cached UTXOs that must be spent within the current block(s).
+    pub num_cached: i32,
```

Comment:
@m-kus instead of maintaining an array of cached outputs, I've added a simple counter that is incremented each time when a cached output is added and decremented when it's spent. So, to ensure all the cached outputs have been spent we just need to check if the counter is zero. This approach is just more efficient.

Please, let me know if it's ok.

Reply:
Neat!